Add support for custom proxy lookup handler

cbornet opened the pull request:
and add a handler that redirects to a URL based on a regex replacement done on the broker URL

nicoloboschi commented on pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/URLRegexLookupProxyHandler.java:
```diff
@@ -0,0 +1,105 @@
+                        performLookup(clientRequestId, lookup.getTopic(), serviceUrl, false, 10)
+                            .whenComplete(
+                                (brokerUrl, ex) -> {
+                                    if (brokerUrl.matches(regex)) {
+                                        String proxyUrl = brokerUrl.replaceAll(regex, replacement);
```
could we optimize here if the proxyUrl is *this* proxy? we could go directly to the broker in this case

cbornet replied:
I'm not sure a proxy knows its URL

nicoloboschi commented on pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/URLRegexLookupProxyHandler.java:
```diff
@@ -0,0 +1,105 @@
+                                    } else {
+                                        proxyConnection.ctx().writeAndFlush(
+                                            Commands.newLookupErrorResponse(ServerError.ServiceNotReady,
+                                                "Broker URL does not match lookup handler regex", clientRequestId));
```
log the brokerUrl for debugging

cbornet replied:
done

nicoloboschi reviewed (commented):
Good work

cbornet committed:
Add support for custom proxy lookup handler
and add a handler that redirects to a URL based on a regex replacement done on the broker URL

eolivelli commented on pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/URLRegexLookupProxyHandler.java:
```diff
@@ -0,0 +1,106 @@
+    @Override
+    public void initialize(ProxyService proxy, ProxyConnection proxyConnection) {
+        super.initialize(proxy, proxyConnection);
+        this.regex = proxy.getConfiguration().getProperties().getProperty("urlRegexLookupProxyHandlerRegex");
```
we should fail here it the property is empty

cbornet replied:
done

eolivelli reviewed (approved):
LGTM

nice work

I would commit the patch on LS in this case and then contribute it upstream when we are sure that everything works as expected in a real environment.
Otherwise we could end up in long discussions with the community.

cbornet committed:
Check that regex parameters are not empty

cbornet commented on pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/URLRegexLookupProxyHandler.java:
```diff
@@ -0,0 +1,106 @@
+        super.initialize(proxy, proxyConnection);
+        this.regex = proxy.getConfiguration().getProperties().getProperty("urlRegexLookupProxyHandlerRegex");
+        this.replacement =
+            proxy.getConfiguration().getProperties().getProperty("urlRegexLookupProxyHandlerReplacement");
```
done

cbornet commented on pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java:
```diff
@@ -329,7 +330,18 @@ private synchronized void completeConnect(AuthData clientData) throws PulsarClie
+                lookupProxyHandler.initialize(service, this);
+            } catch (Exception e) {
+                LOG.error("Failed to initialize lookup proxy handler", e);
+                ctx().writeAndFlush(Commands.newError(-1, ServerError.ServiceNotReady,
```
I'm not sure which server error code to use here.

cbornet commented on pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/URLRegexLookupProxyHandler.java:
```diff
@@ -0,0 +1,113 @@
+                                                false));
+                                    } else {
+                                        proxyConnection.ctx().writeAndFlush(
+                                            Commands.newLookupErrorResponse(ServerError.ServiceNotReady,
```
I'm not sure which server error code to use here.

nicoloboschi commented on pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/URLRegexLookupProxyHandler.java:
```diff
@@ -0,0 +1,113 @@
+                        performLookup(clientRequestId, lookup.getTopic(), serviceUrl, false, 10)
+                            .whenComplete(
+                                (brokerUrl, ex) -> {
+                                    if (brokerUrl.matches(regex)) {
```
we need to catch IAE in case the regex is malconfigured otherwise no error is displayed and the connection just times out

cbornet replied:
done

nicoloboschi reviewed (commented):
I'd add some debug lines for helping users doing the setup (regex exps are hard) 
for example:
>2023-02-27T08:59:37,891+0000 [pulsar-proxy-io-2-4] INFO  org.apache.pulsar.proxy.server.URLRegexLookupProxyHandler - check if matches with value: pulsar://pulsar-broker-az2-0.pulsar-broker-az2.ns.svc.cluster.local:6650                                                                                                                                             │
│ 2023-02-27T08:59:37,891+0000 [pulsar-proxy-io-2-4] INFO  org.apache.pulsar.proxy.server.URLRegexLookupProxyHandler - matched, new value is pulsar://pulsar-proxy-az2:6650

cbornet commented:
> I'd add some debug lines for helping users doing the setup (regex exps are hard)

Should they be INFO or DEBUG level ? (the log may print for every client first connection)

cbornet committed:
Handle regex exceptions and add debug logs

nicoloboschi commented:
@cbornet another quick improvement is to cache the Pattern in the constructor. 
```
Pattern p = Pattern.compile(regex);
```

cbornet committed:
Cache regex Pattern to improve performance

cbornet commented:
👍 
done.

lhotari commented:
@nicoloboschi @cbornet Are we planning to upstream this?

lhotari commented:
@mukesh-ctds @srinath-ctds @nikhil-ctds @cbornet @nicoloboschi This would be a great improvement to Apache Pulsar. Do you have plans to upstream this change? This PR came up in https://github.com/apache/pulsar/issues/23963#issuecomment-2655894604.
